src/algebra/algebra/non_unital.lean

```diff
@@ -0,0 +1,170 @@
+
+  * `non_unital_non_assoc_algebra`
+  * `non_unital_non_assoc_algebra_hom`
+
+## Tags
+
+non-unital, non-associative, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+/-- A not-necessarily-unital, not-necessarily-associative algebra. -/
+@[protect_proj]
+class non_unital_non_assoc_algebra [non_unital_non_assoc_semiring A] extends semimodule R A :=
+(smul_mul_assoc' : ∀ (t : R) (a b : A), (t • a) * b = t • (a * b))
+(mul_smul_comm' : ∀ (t : R) (a b : A), a * (t • b) = t • (a * b))
+
+@[priority 200] -- see Note [lower instance priority]
+instance semiring.to_non_unital_non_assoc_semiring [semiring A] : non_unital_non_assoc_semiring A :=
+{ ..(infer_instance : semiring A) }
+
+@[priority 200] -- see Note [lower instance priority]
+instance algebra.to_non_unital_non_assoc_algebra (R : Type u₁)
+  [comm_semiring R] [semiring A] [algebra R A] : non_unital_non_assoc_algebra R A :=
+{ smul_mul_assoc' := algebra.smul_mul_assoc,
+  mul_smul_comm'  := algebra.mul_smul_comm,
+  ..algebra.to_semimodule }
+
+namespace non_unital_non_assoc_algebra
+
+variables {R A} [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+
+lemma smul_mul_assoc (t : R) (a b : A) :
+  (t • a) * b = t • (a * b) :=
+non_unital_non_assoc_algebra.smul_mul_assoc' t a b
+
+lemma mul_smul_comm (t : R) (a b : A) :
+  a * (t • b) = t • (a * b) :=
+non_unital_non_assoc_algebra.mul_smul_comm' t a b
+
+/-- The algebra multiplication as a bilinear map. -/
+def lmul {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A] :
+  A →ₗ[R] module.End R A :=
+linear_map.mk₂ R (*) add_mul smul_mul_assoc mul_add mul_smul_comm
+
+/-- If the underlying `non_unital_non_assoc_semiring` is actually a `semiring` we have an
+`algebra`. -/
+def to_algebra {R : Type u₁} {A : Type u₂}
+  [comm_semiring R] [semiring A] [non_unital_non_assoc_algebra R A] : algebra R A :=
+algebra.of_semimodule smul_mul_assoc mul_smul_comm
+
+end non_unital_non_assoc_algebra
+
+variables [non_unital_non_assoc_semiring A] [non_unital_non_assoc_algebra R A]
+variables [non_unital_non_assoc_semiring B] [non_unital_non_assoc_algebra R B]
+variables [non_unital_non_assoc_semiring C] [non_unital_non_assoc_algebra R C]
+
+set_option old_structure_cmd true
+
+/-- A morphism of non-associative algebras. -/
+structure non_unital_non_assoc_algebra_hom extends A →ₗ[R] B, mul_hom A B
+
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_linear_map
+attribute [nolint doc_blame] non_unital_non_assoc_algebra_hom.to_mul_hom
+
+namespace non_unital_non_assoc_algebra_hom
+
+variables {R A B C}
+
+instance {R A B} [comm_semiring R] [ring A] [ring B] [algebra R A] [algebra R B] :
+  has_coe (A →ₐ[R] B) (non_unital_non_assoc_algebra_hom R A B) :=
+⟨λ f, { to_fun    := f,
+        map_add'  := f.map_add,
+        map_smul' := f.map_smul,
+        map_mul'  := f.map_mul, }⟩
```

Comment:
I suspect (but haven't tested) that this works:
```suggestion
⟨λ f, { to_fun    := f, ..f }⟩
```

Reply:
Unfortunately there is no `alg_hom.map_smul'` but we can get half way there so I'll commit what we can.